Source/KSCrash/Source/KSCrash/Recording/Tools/BSG_KSMachHeaders.m

```diff
@@ -0,0 +1,173 @@
+    
+    // Early exit conditions; this is not a valid/useful binary image
+    // 1. We can't find a sensible Mach command
+    uintptr_t cmdPtr = bsg_ksdlfirstCmdAfterHeader(header);
+    if (cmdPtr == 0) {
+        return false;
+    }
+
+    // 2. The image doesn't have a name.  Note: running with a debugger attached causes this condition to match.
+    Dl_info DlInfo = (const Dl_info) { 0 };
+    dladdr(header, &DlInfo);
+    const char *imageName = DlInfo.dli_fname;
+    if (!imageName) {
+        return false;
+    }
+    
+    // Look for the TEXT segment to get the image size.
+    // Also look for a UUID command.
+    uint64_t imageSize = 0;
+    uint64_t imageVmAddr = 0;
+    uint8_t *uuid = NULL;
+
+    for (uint32_t iCmd = 0; iCmd < header->ncmds; iCmd++) {
+        struct load_command *loadCmd = (struct load_command *)cmdPtr;
+        switch (loadCmd->cmd) {
+        case LC_SEGMENT: {
+            struct segment_command *segCmd = (struct segment_command *)cmdPtr;
+            if (strcmp(segCmd->segname, SEG_TEXT) == 0) {
+                imageSize = segCmd->vmsize;
+                imageVmAddr = segCmd->vmaddr;
+            }
+            break;
+        }
+        case LC_SEGMENT_64: {
+            struct segment_command_64 *segCmd =
+                (struct segment_command_64 *)cmdPtr;
+            if (strcmp(segCmd->segname, SEG_TEXT) == 0) {
+                imageSize = segCmd->vmsize;
+                imageVmAddr = segCmd->vmaddr;
+            }
+            break;
+        }
+        case LC_UUID: {
+            struct uuid_command *uuidCmd = (struct uuid_command *)cmdPtr;
+            uuid = uuidCmd->uuid;
+            break;
+        }
+        }
+        cmdPtr += loadCmd->cmdsize;
+    }
+    
+    // Save these values
+    info->mh = header;
+    info->cpusubtype = header->cpusubtype;
+    info->cputype = header->cputype;
+    info->imageSize = imageSize;
+    info->imageVmAddr = imageVmAddr;
+    info->uuid = uuid;
+    info->name = imageName;
+    
+    return true;
+}
+
+/**
+ * A callback invoked when dyld loads binary images.  It stores enough relevant info about the
+ * image to populate a crash report later.
+ *
+ * @param header A mach_header structure
+ *
+ * @param slide A virtual memory slide amount. The virtual memory slide amount specifies the difference between the
+ *              address at which the image was linked and the address at which the image is loaded.  Unused.
+ */
+void bsg_mach_binary_image_added(const struct mach_header *header, intptr_t slide)
+{
+    BSG_Mach_Binary_Image_Info info = { 0 };
+    if (bsg_populate_mach_image_info(header, &info)) {
+        bsg_add_mach_binary_image(info);
+    }
+}
+
+/**
+ * Called when a binary image is unloaded.
+ */
+void bsg_mach_binary_image_removed(const struct mach_header *header, intptr_t slide)
+{
+    // Convert header and slide into an info struct
+    BSG_Mach_Binary_Image_Info info;
```

Comment:
Consider initializing the struct in the same manner as `bsg_mach_binary_image_added` for consistency